main/src/ca/uwaterloo/flix/language/phase/Desugar.scala

```diff
@@ -0,0 +1,145 @@
+      case (k, v) => visitUnit(k, v)
+    }
+
+    val desugaredCompilationUnits = unitsVal.foldLeft(Map.empty[Ast.Source, DesugaredAst.CompilationUnit]) {
+      case (macc, (k, v)) => macc + (k -> v)
+    }
+
+    DesugaredAst.Root(desugaredCompilationUnits, program.entryPoint, program.names)
+  }
+
+  /**
+    * Desugars the given compilation unit `unit`.
+    */
+  private def visitUnit(src: Ast.Source, unit: WeededAst.CompilationUnit)(implicit flix: Flix): (Ast.Source, DesugaredAst.CompilationUnit) = unit match {
+    case WeededAst.CompilationUnit(usesAndImports0, decls0, loc) =>
+      val usesAndImports = usesAndImports0.map(visitUseOrImport)
+      val decls = decls0.map(visitDecl)
+      src -> DesugaredAst.CompilationUnit(usesAndImports, decls, loc)
+  }
+
+  /**
+    * Maps `u0` to a corresponding [[DesugaredAst.UseOrImport]].
+    */
+  private def visitUseOrImport(u0: WeededAst.UseOrImport): DesugaredAst.UseOrImport = u0 match {
+    case WeededAst.UseOrImport.Use(qname, alias, loc) => DesugaredAst.UseOrImport.Use(qname, alias, loc)
+    case WeededAst.UseOrImport.Import(name, alias, loc) => DesugaredAst.UseOrImport.Import(name, alias, loc)
+  }
+
+  /**
+    * Compiles `decl` to a list of [[DesugaredAst.Declaration]].
+    */
+  private def visitDecl(decl: WeededAst.Declaration)(implicit flix: Flix): DesugaredAst.Declaration = decl match {
+    case WeededAst.Declaration.Namespace(ident, usesAndImports, decls, loc) =>
+      val usesAndImportsVal = usesAndImports.map(visitUseOrImport)
+      val declsVal = decls.map(visitDecl)
+      DesugaredAst.Declaration.Namespace(ident, usesAndImportsVal, declsVal, loc)
+
+    case d: WeededAst.Declaration.Class => visitClass(d)
+    case d: WeededAst.Declaration.Instance => visitInstance(d)
+    case d: WeededAst.Declaration.Def => visitDef(d)
+    case d: WeededAst.Declaration.Law => visitLaw(d)
+    case d: WeededAst.Declaration.Enum => visitEnum(d)
+    case d: WeededAst.Declaration.RestrictableEnum => visitRestrictableEnum(d)
+    case d: WeededAst.Declaration.TypeAlias => visitTypeAlias(d)
+    case d: WeededAst.Declaration.Effect => visitEffect(d)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.Declaration.Class]] `d`.
+    */
+  private def visitClass(d: WeededAst.Declaration.Class)(implicit flix: Flix): DesugaredAst.Declaration.Class = ???
+
+  /**
+    * Desugars the given [[WeededAst.Declaration.Instance]] `d`.
+    */
+  private def visitInstance(d: WeededAst.Declaration.Instance)(implicit flix: Flix): DesugaredAst.Declaration.Instance = ???
+
+  /**
+    * Desugars the given [[WeededAst.Declaration.Def]] `d`.
+    */
+  private def visitDef(d: WeededAst.Declaration.Def)(implicit flix: Flix): DesugaredAst.Declaration.Def = d match {
+    case WeededAst.Declaration.Def(doc, ann, mod, ident, tparams, fparams, exp, tpe, eff, tconstrs, constrs, loc) =>
+      flix.subtask(ident.name, sample = true)
+
+      val tparamsVal = visitKindedTypeParams(tparams)
```

Comment:
I guess maybe we should drop the "Val" part of the name, since these are not Validations.

Reply:
Ah, okay, I thought it was short for “value”.